README.md
[![Build Status](https://travis-ci.org/ifwe/monocle-api.png)](https://travis-ci.org/ifwe/monocle-api)

Monocle API Router for Connect
==============================

Monocle is a an API router that focuses on *consistency*, *flexibility* and *performance*.

## Consistency

Monocle implements various features to encourage consistency across all endpoints.

### Schemas

Monocle requires that each resource path includes a schema defined in [JSON Schema](http://json-schema.org/) format. The schema is shared among all HTTP verbs for the resource, e.g. `GET`, `POST`, `PUT`, etc. Monocle also validates all input and output with the schema, ensuring that the schema remains the source of truth for how data is structured.

### Server-Side Symlinks

Monocle supports "Server-Side Symlinks" to easily embed a resource in another resource. The Monocle Framework will automatically resolve these symlinks on the server, resulting in a complete response for clients. By utilizing Server-Side Symlinks, consistency comes for free when the Symlinked resources are updated.

## Flexibility

Monocle makes no assumptions about how clients consume the data -- the only assumption it makes is that client needs will change regularly.

### Property Filtering

Monocle allows clients to specify how much data to return for any given resource. Property filtering supports top-level properties, nested properties, and plucking properties from arrays. See [Monocle API Props](https://github.com/ifwe/monocle-api-props) for details on how property filtering works.

## Performance

Monocle allows for performance optimization via caching and property routing.

### Caching

Monocle provides cache-control headers for basic REST, but also supports nested resource caching when using the [Monocle Client](https://github.com/ifwe/monocle-client-js). Collection caching is also supported via the use of weak e-tags.

### Property Routing

For large resources, Monocle allows you to provide it with various callbacks that map to each of the properties in the resource. When used with *Property Filtering*, Monocle will automatically call only the required callbacks to fulfill the request.

--

## Quick Start

To get started, make sure you're using *Node 8.8.1* or greater, and create a new node project in an empty directory:

```
$ mkdir my-monocle-server
$ cd my-monocle-server
$ npm init
```

After filling in the details, install the required dependencies:

```
$ npm install monocle-api connect body-parser

```

Create a file named `server.js` and add the following code:


```js
const connect = require('connect');
const app = connect();

// Allow parsing of JSON-encoded request body
const bodyParser = require('body-parser');
app.use(bodyParser.json());

// Create an API Router instance
const MonocleApi = require('monocle-api');
const api = new MonocleApi();
const Resource = MonocleApi.Resource;

// For this simple demo we'll set up a simple in-memory data store for the user resources.
const users = {
  1: {
    userId: 1,
    displayName: 'Alice',
    age: 27,
    gender: 'female',
  },
  2: {
    userId: 2,
    displayName: 'Fred',
    age: 22,
    gender: 'male',
  },
};

// Configure your first API route
api.route(
  // Define the URL pattern for this resource
  '/users/:userId',

  // Define the schema for this resource. The schema will be shared across the supported HTTP methods for this resource.
  {
    name: 'User',
    description: 'A user resource',
    type: 'object',
    properties: {
      userId: { type: 'integer', minimum: 1, readOnly: true, sample: 123 },
      displayName: { type: 'string', minLength: 1, maxLength: 255 },
      age: { type: 'integer', minimum: 18, maximum: 99 },
      gender: { type: 'string', enum: ['male', 'female'] },
    },
  },

  // Define the HTTP methods that are supported by this url.
  {
    // Handle GET requests for this resource
    get: (request) => {
      let userId = request.getParam('userId'); // extracts userId param from url, automatically casts it to int due to schema definition

      let user = users[userId];
      if (!user) {
        return request.error(404, { message: 'User not found' });
      }

      // Resolve promise with the user object and it will be converted to JSON automatically
      // Monocle will also validate the return value and return a 500-level error code if the value does not validate.
      return new Resource(`/users/${userId}`, user, 60000);
    },

    // Handle PUT requests for this resource
    put: (request) => {
      let userId = request.getParam('userId'); // extracts userId param from url, automatically casts it to int due to schema definition

      // Replace entire user object with provided resource, which is automatically JSON-decoded
      // Monocle will have rejected this request if the provided resource did not validate with the schema.
      user = request.getResource();

      // Resolve promise with the updated user object
      return new Resource(`/users/${userId}`, user, 60000);
    },
  }
);

// Add the API middleware to your connect app
app.use(api.middleware());

// Create web server and listen on port 5150
const http = require('http');
http.createServer(app).listen(5150, function() {
  console.log("Monocle API is now listening on port 5150");
});
```

You can now start your monocle server by running `node server.js`.

## REST

Monocle API supports RESTful API calls:

```bash
$ curl -i http://127.0.0.1:5150/users/1
HTTP/1.1 200 OK
Content-Type: application/json
cache-control: private, max-age=60000
Date: Thu, 08 Feb 2018 23:27:30 GMT
Connection: keep-alive
Content-Length: 144

{
  "$type": "resource",
  "$id": "/users/1",
  "$expires": 60000,
  "userId": 1,
  "displayName": "Alice",
  "age": 27,
  "gender": "female"
}


$ curl -X PUT -H 'Content-Type: application/json' -d '{"displayName": "Joe", "age": 42, "gender": "male"}' -i http://127.0.0.1:5150/users/1
HTTP/1.1 200 OK
Content-Type: application/json
cache-control: private, max-age=60000
Date: Thu, 08 Feb 2018 23:29:48 GMT
Connection: keep-alive
Content-Length: 125

{
  "$type": "resource",
  "$id": "/users/1",
  "$expires": 60000,
  "displayName": "Joe",
  "age": 42,
  "gender": "male"
}


$ curl -i http://127.0.0.1:5150/users/3
HTTP/1.1 404 Not Found
Content-Type: application/json
Date: Thu, 08 Feb 2018 23:37:36 GMT
Connection: keep-alive
Content-Length: 145

{
  "code": 2000,
  "error": "UNKNOWN",
  "message": "User not found",
  "properties": [],
  "$httpStatus": 404,
  "$httpMessage": "NOT FOUND"
}

```

Property filtering can be used to restrict how much data is returned.

```bash
$ curl -i http://127.0.0.1:5150/users/1?props=displayName,age
HTTP/1.1 200 OK
Content-Type: application/json
cache-control: private, max-age=60000
Date: Thu, 08 Feb 2018 23:39:04 GMT
Connection: keep-alive
Content-Length: 107

{
  "$type": "resource",
  "$id": "/users/1",
  "$expires": 60000,
  "displayName": "Alice",
  "age": 27
}

```

See `demo/index.js` for advanced usage.

## Files and Directory Structure

The following describes the various files in this repo and the directory structure.

**Note:** Files and directories prefixed by `*` are auto-generated and excluded from the
repository via `.gitignore`.

    .
    ├── Gruntfile.js            # grunt task configuration
    ├── README.md               # this file
    ├── *docs                   # autogenerated documentation
    │   └── *index.html         # each JS file in `./lib` has a corresponding HTML file for documentation
    ├── lib                     # all code for this library will be placed here
    │   └── index.js            # main entry point for the API router
    ├── *node_modules           # all dependencies will be installed here by npm
    ├── package.json            # description of this package for npm, including dependency lists
    └── test                    # unit test configuration, reports, and specs
        ├── *coverage.html      # code coverage report
        ├── lib                 # specs go here with a 1:1 mapping to code in `./lib`
        │   └── index_test.js   # spec for `./lib/index.js`
        ├── mocha.opts          # runtime options for mocha
        └── test_runner.js      # configures mocha environment (e.g. chai, sinon, etc.)

## Development

### Grunt

Grunt is a JavaScript task runner to automate common actions. The API Router project
supports the following grunt tasks:

**test**

Runs all unit tests through mocha.

    $ grunt test

**coverage**

Runs all unit tests and generates a code coverage report in `./test/coverage.html`

    $ grunt coverage

**watch**

Automatically runs mocha tests each time a file changes in `./lib` or `./test`.

    $ grunt watch

**docs**

Generates documentation for all JS files within `./lib` using docco. Documentation is
written to `./docs`.

    $ grunt docs

**clean**

Deletes all auto-generated files, including `./docs` and `./test/coverage.html`

### Mocha, Sinon, Chai, Blanket

The ultimate TDD environment for node. Place your specs in `./test/lib`, and run `grunt test`.

See `./test/lib/index_test.js` for examples.


CHANGELOG.md
3.2.0
=====

+ Drop support for Node.js < 12 and update supported versions in "engines" in package.json
+ Hold back jsen dependency. See https://github.com/ifwe/monocle-api/issues/86
+ Drop dependency on bluebird - it recommends using native Promises when they're available.

3.1.0
=====

+ Update dependencies, excluding `jsen`. See https://github.com/ifwe/monocle-api/issues/86


demo/index.js
// This is an example route configuration file.
// This file manages two related resources:
// - /users/123     - Single user
// - /users         - Collection of users

/*** Initial dependencies **/

var _ = require('lodash');
var Router = require('../lib');
var Resource = require('../lib').Resource;
var OffsetPaginator = require('../lib').OffsetPaginator;
var Symlink = require('../lib').Symlink;
var fs = require('fs');
var path = require('path');
var chain = require('connect-chain');

/*** Set up simple HTTP server ***/

var connect = require('connect');
var app = connect();

// Allow method override via ?_method=METHOD query string parameter
var methodOverride = require('method-override');
app.use(methodOverride('_method', {
    methods: [ 'GET', 'POST' ] // Specifies which methods can support overrides
}));
var bodyParser = require('body-parser');
app.use(bodyParser.json());

/*** Mock data ***/

var mockUserBasicInfo = require('./data/user-basic-info');
var mockUserEmails = require('./data/user-emails');
var mockPhotos = require('./data/photos');

/*** Define the schemas ***/

var userSchema = require('./schemas/user');
var userCollectionSchema = require('./schemas/user-collection');
var photoSchema = require('./schemas/photo');
var photoCollectionSchema = require('./schemas/photo-collection');

/*** Set up a new router ***/

var api = new Router({
    "openapi": "3.0.0",
    "info": {
        "version": 1,
        "description": "Some Demo Stuff",
        "title": "Demo App"
    }
});

/*** Define the routes ***/

// Define the route that points to a specific user.
// The URL may contain named parameters, which will be parsed and handed to the callback.
api.route('/users/:userId', userSchema, {
    // Complex resources may need multiple callback handlers to support different properties.
    // The API router will figure out which callbacks are necessary to satisfy the incoming request.
    get: [
        {
            props: ['displayName', 'age', 'gender', 'birthDate', 'city', 'country'],
            callback: getUserBasicInfo
        },
        {
            props: ['email'],
            callback: getUserEmailInfo
        }
    ],
    put: [
        {
            props: ['displayName', 'age', 'gender', 'birthDate', 'city', 'country'],
            callback: putUserBasicInfo
        },
        {
            props: ['email'],
            callback: putUserEmailInfo
        }
    ],
    patch: function() {},
    delete: function(request) {
        var userId = request.getParam('userId');
        return request.error(403, 'You are not allowed to delete this user.');
    }
});

// Defines the route that manages a collection of users
api.route(['/users', 'limit=10&offset=0&search'], userCollectionSchema, {
    get: getUsers,
    post: createUser
});

api.route(['/users/:userId/photos', 'limit=10&offset=0'], photoCollectionSchema, {
    // Complex resources may need multiple callback handlers to support different properties.
    // The API router will figure out which callbacks are necessary to satisfy the incoming request.
    get: function(request) {
        var userId = request.getParam('userId');

        //Get Parametersplice(0, limit).
        var limit = request.getQuery('limit');
        var offset = request.getQuery('offset');

        var photos = mockPhotos.filter(function(photo) {
            return photo.userId === userId;
        })
        .slice(offset, limit)
        .map(function(photo) {
            return new Resource('/users/' + userId + '/photos/' + photo.photoId, photo, 86400);
        });

        return new OffsetPaginator('/users/' + userId + '/photos')
        .setItems(photos)
        .setLimit(limit)
        .setOffset(offset)
        .setTotal(mockPhotos.length);
    },
    post: function(request) {
        return request.getUpload('photo')
        .then(function(upload) {
            var photoFileName = Math.round(Math.random() * 1e16).toString(36);

            switch (upload.mimeType) {
                case 'image/png':
                    photoFileName += '.png';
                    break;

                case 'image/jpeg':
                    photoFileName += '.jpg';
                    break;

                case 'image/gif':
                    photoFileName += '.gif';
                    break;

                default:
                    return request.error(442, 'Invalid upload type: ' + upload.mimeType);
            }

            var saveFilePath = path.join(__dirname, 'files', 'photos', photoFileName);

            var wstream = fs.createWriteStream(saveFilePath, {
                encoding: upload.encoding
            });

            wstream.write(upload.buffer);
            wstream.end();

            var userId = request.getParam('userId');
            var photoId = mockPhotos.length;
            var photo = request.getResource();
            photo.userId = userId;
            photo.photoId = photoId;
            photo.url = 'http://localhost:5150/photos/' + photoFileName;
            mockPhotos.push(photo);

            return new Resource('/users/' + userId + '/photos/' + photo.photoId, photo, 86400);
        });
    }
});

api.route('/users/:userId/photos/:photoId', photoSchema, {
    // Complex resources may need multiple callback handlers to support different properties.
    // The API router will figure out which callbacks are necessary to satisfy the incoming request.
    get: function(request) {
        var userId = request.getParam('userId');
        var photoId = request.getParam('photoId');
        for (var i = 0, len = mockPhotos.length; i < len; i++) {
            var photo = mockPhotos[i];
            if (photo.photoId === photoId && photo.userId === userId) {
                return new Resource('/users/' + userId + '/photos/' + photoId, photo, 86400);
            }
        }
        return Promise.reject("Unable to find photo for user " + userId + " and photo id " + photoId);
    },
    put: function(request) {
        var userId = request.getParam('userId');
        var photoId = request.getParam('photoId');
        var submittedPhoto = request.getResource();
        var photo;
        for (var i = 0, len = mockPhotos.length; i < len; i++) {
            if (mockPhotos[i].photoId === photoId && mockPhotos[i].userId === userId) {
                photo = mockPhotos[i];
                break;
            }
        }
        if (!photo) {
            return Promise.reject("Unable to find photo for user " + userId + " and photo id " + photoId);
        }
        photo.caption = submittedPhoto.caption;
        // mockPhotos[i] = photo;
        return new Resource('/users/' + userId + '/photos/' + photo.photoId, photo, 86400);
    },
    delete: function() {},
});

/*** Define the callback functions ***/

// Returns basic user info for a given user id
function getUserBasicInfo(request) {
    // Get the user ID from the URL params
    var userId = request.getParam('userId'); // ex: 123

    return new Promise(function(resolve, reject) {
        for (var i = 0, len = mockUserBasicInfo.length; i < len; i++) {
            if (mockUserBasicInfo[i].userId === userId) {
                return resolve(new Resource('/users/' + userId, mockUserBasicInfo[i], 3600));
            }
        }
        reject('Unable to find basic info for user id ' + userId);
    });
}

// Returns email info for a given user id
function getUserEmailInfo(request) {
    // Get the user ID from the URL params
    var userId = request.getParam('userId'); // ex: 123

    return new Promise(function(resolve, reject) {
        var index = userId - 1;
        var userEmail = mockUserEmails[index];
        if (userEmail) {
            resolve(new Resource('/users/' + userId, userEmail, 3600));
        } else {
            reject('Unable to find email for user id ' + userId);
        }
    });
}

// Returns basic info for a collection of users
function getUsersBasicInfo(request) {
    return new Promise(function(resolve, reject) {
        resolve(mockUserBasicInfo.map(function(info, i) {
            return new Resource('/users/' + info.userId, info);
        }));
    });
}

// Returns email info for a collection of users
function getUsersEmailInfo(request) {
    return new Promise(function(resolve, reject) {
        resolve(mockUserEmails.map(function(info, i) {
            return new Resource('/users/' + info.userId, info);
        }));
    });
}

// Updates a user
function putUserBasicInfo(request) {
    var userId = request.getParam('userId');
    var resource = request.getResource();

    return new Promise(function(resolve, reject) {
        for (var i = 0, len = mockUserBasicInfo.length; i < len; i++) {
            if (mockUserBasicInfo[i].userId === userId) {
                for (var k in resource) {
                    mockUserBasicInfo[i][k] = resource[k];
                }
                return resolve(new Resource('/users/' + userId, mockUserBasicInfo[i], 3600));
            }
        }
        reject('Unable to find basic info for user id ' + userId);
    });
}

function putUserEmailInfo(request) {
    var userId = request.getParam('userId');
    var resource = request.getResource();
    return new Resource('/users/' + userId, resource, 3600);
}

// Gets all users
function getUsers(request, connection) {
    // The limit is guaranteed to be an integer and within range, or default value
    var limit = request.getQuery('limit');

    // The offset is guaranteed to be an integer and within range, or default value
    var offset = request.getQuery('offset');

    var search = request.getQuery('search');

    var users = mockUserBasicInfo.slice(0); // Make a copy of users
    if (search) {
        search = search.toLowerCase();
        users = users.filter(function(user) {
            return (-1 !== user.displayName.toLowerCase().indexOf(search));
        });
    }

    var users = users.slice(offset, offset + limit).map(function(user) {
        return new Symlink('/users/' + user.userId);
    });

    return new OffsetPaginator('/users')
    .setItems(users)
    .setExpires(1000)
    .setLimit(limit)
    .setOffset(offset);
}

// Create a user
function createUser(request) {

}

// A very simple photo upload page
app.use(function(req, res, next) {
    if (req.url !== '/upload-photo' || req.method !== 'GET') {
        return next();
    }
    res.setHeader('Content-Type', 'text/html');
    res.write('\
        <form method="POST" enctype="multipart/form-data" action="/my-api/users/1/photos">\
            <input type="file" name="photo" />\
            <textarea name="caption" placeholder="caption"></textarea>\
            <button>Upload</button>\
        </form>\
    ');
    res.end();
});

/*** Set up some aliases ***/
api.alias('/alice', '/users/1');

/*** Set up some aliases ***/
api.alias('/random', function(request, connection) {
    var userId = Math.ceil(Math.random() * mockUserBasicInfo.length);
    request.setResourceId('/users/' + userId);
    return request;
});

/*** Set up some aliases ***/
api.alias('/me', function(request, connection) {
    request.setResourceId('/users/2');
    return request;
});

// Allow uploaded files to be served
var serveStatic = require('serve-static');
app.use(serveStatic(path.join(__dirname, 'files')));

var swagger = require('swagger-ui-dist').getAbsoluteFSPath();
app.use("/swagger", chain(
  function(req, res, next) {
      if(req.method === "GET" && req.url === '/') {
          res.writeHead(302, {'location': '?url=http://localhost:5150/my-api'});
          return res.end();
      }

      next();
  },
  serveStatic(swagger)
));

// Mount the API as middleware
app.use(api.middleware({
    basePath: '/my-api'
}));

api.on('api:success', function(data) {
    console.log('Success!!', data.resourceId);
});

api.on('api:error', function(data) {
    console.log('Error :(', data.resourceId);
});

// Create web server and listen on port 5150
var http = require('http');
http.createServer(app).listen(5150, function() {
    console.log("API Router Demo listening on port 5150");
    console.log("View documentation: http://localhost:5150/my-api");
});
